### PR TITLE
fix: fire close on fail for WebSocket

### LINKED ISSUE #3546: WebSockets do not fire 'close' event if the connection failed to be established
Consider the following code:
```js
const webSocket = new undici.WebSocket("wss://invalid-domain.example.com/");
webSocket.onopen = () => { console.log("open"); };
webSocket.onclose = () => { console.log("close"); };
webSocket.onerror = () => { console.log("error"); };
```
It outputs:
```
error
```

However, a standard-compliant WebSocket implementation would output:
```
error
close
```

This is because [establish a WebSocket connection](https://websockets.spec.whatwg.org/#concept-websocket-establish) algorithm would invoke [fail the WebSocket Connection](https://datatracker.ietf.org/doc/html/rfc6455#section-7.1.7) algorithm in step 11.1, which would [Close the WebSocket Connection](https://datatracker.ietf.org/doc/html/rfc6455#section-7.1.1). The WebSocket standard states:

> When [the WebSocket connection is closed](https://datatracker.ietf.org/doc/html/rfc6455#section-7.1.4), possibly cleanly, the user agent must [queue a task](https://html.spec.whatwg.org/multipage/webappapis.html#queue-a-task) to run the following substeps:
> ...
> 3. [Fire an event](https://dom.spec.whatwg.org/#concept-event-fire) named close at the [WebSocket](https://websockets.spec.whatwg.org/#websocket) object.

This bug is reproducible on commit [4b8958a](https://github.com/nodejs/undici/commit/4b8958a228aae55366dfba1dc70f03f08bc2d48d).

### PR DESCRIPTION
**Pull request by eXhumer**
## This relates to...

Fixes #3546

## Rationale

`close` event is never fired after a WebSocket connection is failed to establish.

## Changes

* change `WebSocket.#onSocketClose` when trying to access `closingInfo`. `WebSocket.#parser` is `undefined` when connection fails to establish.
* run `WebSocket.#onSocketClose` after firing `error` event, which in turn fires the `close` event.

### Features

_N/A_

### Bug Fixes

* fix `close` event not firing on failed connection.

### Breaking Changes and Deprecations

_N/A_

## Status

- [x] I have read and agreed to the [Developer's Certificate of Origin][cert]
- [x] Tested
- [S] Benchmarked (**optional**)
- [S] Documented
- [x] Review ready
- [x] In review
- [ ] Merge ready

[cert]: https://github.com/nodejs/undici/blob/main/CONTRIBUTING.md

### CODE REVIEW

**Comment by marximimus** [comment]
LGTM

**Review by KhafraDev — CHANGES_REQUESTED** [review]
please add a test

File: lib/web/websocket/websocket.js [hunk 1]
```diff
@@ -666,7 +668,7 @@ class WebSocket extends EventTarget {
     let code = 1005
     let reason = ''
 
-    const result = this.#parser.closingInfo
+    const result = this.#parser?.closingInfo
```
Comment (Uzlopak) [inline lib/web/websocket/websocket.js:671]: why this change?

Comment (eXhumer) [reply lib/web/websocket/websocket.js:671]: `WebSocket.#socketClose` will thrown an error trying to access properties on `undefined` as `WebSocket.#parser` is `undefined` when the connection never establishes.

Comment (Uzlopak) [reply lib/web/websocket/websocket.js:671]: seems odd

**Comment by eXhumer** [comment]
Some of the autobahn test keep failing locally due to tests timing out. If there is any actual issue unrelated to timeout, please let me know.

File: lib/web/websocket/websocket.js (hunk 1 above)
Comment (eXhumer) [reply lib/web/websocket/websocket.js:671]: https://github.com/nodejs/undici/blob/89a46dd54f7d7db7513c435fac8042769ee9e9b5/lib/web/websocket/websocket.js#L481-L490

`WebSocket.#parser` is set inside `WebSocket.#onConnectionEstablished`, which is never reached since an actual connection is never established.

File: test/websocket/issue-3546.js [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+test('close is fired on failed connection', async (t) => {
+  const { completed, deepStrictEqual } = tspl(t, { plan: 1 })
+  const ws = new WebSocket('ws://localhost:1')
+
+  ws.addEventListener('close', () => {
+    deepStrictEqual(ws.readyState, WebSocket.CLOSED)
+  })
+
+  await completed
+})
```
Comment (Uzlopak) [inline test/websocket/issue-3546.js:16]: imho check the right order too

```suggestion
test('first error than close event is fired on failed connection', async (t) => {
  const { completed, strictEqual } = tspl(t, { plan: 4 })
  const ws = new WebSocket('ws://localhost:1')

  let orderOfEvent = 0

  ws.addEventListener('error', () => {
    strictEqual(orderOfEvent++, 0)
    strictEqual(ws.readyState, WebSocket.CLOSED)
  })

  ws.addEventListener('close', () => {
    strictEqual(orderOfEvent++, 1)
    strictEqual(ws.readyState, WebSocket.CLOSED)
  })

  await completed
})
```

**Review by Uzlopak — CHANGES_REQUESTED** [review]
I think the test can test a little bit more

File: lib/web/websocket/websocket.js (hunk 1 above)
Comment (Uzlopak) [reply lib/web/websocket/websocket.js:671]: i checked it and it seems ok.

**Review by Uzlopak — APPROVED** [review]
LGTM

**Review by KhafraDev — CHANGES_REQUESTED** [review]
a bunch of autobahn tests are failing

**Comment by Uzlopak** [comment]
We can remove line 525, where we set the connection status, because we call #onSocketClose.

And the change in #onSocketClose with optional chaining brakes the autobahn. Because it changes the error code to 1006 instead of 1005.

Probably need to make #onSocketClose accept a parameter. If we pass a symbol kFail than we handle the onFail case and code stays at 1005. Or maybe use the wasClean variable?

**Review by Uzlopak — CHANGES_REQUESTED** [review]
.

**Comment by KhafraDev** [comment]
The issue is that onSocketClose is not appropriate to call here. You need to handle setting the state to CLOSED and emitting a close event in a separate function that both onFail and onSocketClose can use.

**Comment by eXhumer** [comment]
Sorry for not getting to back to this earlier. After KhafkaDev said there were autobahn tests failing, I wanted to properly setup the tests locally to make sure they were working as expected. Although the test seem to be running fine on CI, it seems to fail test cases between 12.4.3 and 12.4.18 locally, while passing on CI. Does anyone have any insight on why this may be the case? And in some cases, re-running the same test locally seem to produce different fails. I have tried to do the autobahn test on two different machines (macOS 15.1 & Windows 11), both seem to fail these 16 tests locally.

```
│ 12.4.3  │ 'FAILED'        │ 'UNCLEAN'       │ 399      │ null            │
│ 12.4.4  │ 'FAILED'        │ 'UNCLEAN'       │ 135      │ null            │
│ 12.4.5  │ 'FAILED'        │ 'UNCLEAN'       │ 91       │ null            │
│ 12.4.6  │ 'FAILED'        │ 'UNCLEAN'       │ 54       │ null            │
│ 12.4.7  │ 'FAILED'        │ 'UNCLEAN'       │ 181      │ null            │
│ 12.4.8  │ 'FAILED'        │ 'UNCLEAN'       │ 351      │ null            │
│ 12.4.9  │ 'FAILED'        │ 'UNCLEAN'       │ 630      │ null            │
│ 12.4.10 │ 'FAILED'        │ 'UNCLEAN'       │ 1022     │ null            │
│ 12.4.11 │ 'FAILED'        │ 'UNCLEAN'       │ 61       │ null            │
│ 12.4.12 │ 'FAILED'        │ 'UNCLEAN'       │ 209      │ null            │
│ 12.4.13 │ 'FAILED'        │ 'UNCLEAN'       │ 419      │ null            │
│ 12.4.14 │ 'FAILED'        │ 'UNCLEAN'       │ 722      │ null            │
│ 12.4.15 │ 'FAILED'        │ 'UNCLEAN'       │ 1147     │ null            │
│ 12.4.16 │ 'FAILED'        │ 'UNCLEAN'       │ 1066     │ null            │
│ 12.4.17 │ 'FAILED'        │ 'UNCLEAN'       │ 1052     │ null            │
│ 12.4.18 │ 'FAILED'        │ 'UNCLEAN'       │ 1042     │ null            │
```

I was unable to do any testing locally as a result.
